guides/schema/code_reuse.md

```diff
@@ -4,76 +4,84 @@ title: Schema — Code Reuse
 
 Besides reducing duplicate code, these approaches also allow you to test parts of your schema in isolation.
 
-## Resolver classes
+## Functions
 
-(Originally described by AlphaSights at https://m.alphasights.com/graphql-ruby-clean-up-your-query-type-d7ab05a47084.)
+{{ "GraphQL::Function" | api_doc }} can hold generalized field logic which can be specialized on a field-by-field basis.
 
-Many examples use a Proc literal for a field's `resolve`, for example:
+To define a function, make a class that extends {{ "GraphQL::Function" | api_doc }}:
 
 ```ruby
-field :name, types.String do
-  # Resolve taking a Proc literal:
-  resolve ->(obj, args, ctx) { obj.name }
-end
-```
+class FindRecord < GraphQL::Function
+  attr_reader :type
 
-However, you can pass _any_ object to `resolve(...)`, so long as it responds to `.call(obj, args, ctx)`.
+  def initialize(model_class:, type:)
+    @model_class = model_class
+    @type = type
+  end
 
-This means you can pass in a module:
+  argument :id, !types.ID
 
-```ruby
-module NameResolver
-  def self.call(obj, args, ctx)
-    obj.name
+  def call(obj, args, ctx)
+    @model_class.find(args[:id])
   end
 end
-
-# use `NameResolver` instead of a proc
-field :name, types.String do
-  resolve(NameResolver)
-end
-# `resolve: resolver` is equivalent to `do resolve(resolver) end`
-# so this is the same:
-field :name, types.String, resolve: NameResolver
 ```
 
-Or, you can pass an instance of a class:
+Then, connect the function to field definitions with the `function:` keyword:
 
 ```ruby
-class MethodCallResolver
-  def initialize(method_name)
-    @method_name = method_name
-  end
+field :product, function: FindRecord.new(model_class: Product, type: Types::ProductType)
+field :category, function: FindRecord.new(model_class: Category, type: Types::CategoryType)
+```
 
-  def call(obj, args, ctx)
-    obj.public_send(@method_name)
+- `#arguments => Hash<String => GraphQL::Argument>`
+- `#type => GraphQL::BaseType`
+- `#call(obj, args, ctx) => Object`
+- `#complexity => Integer, Proc`
+- `#description => String, nil`
+- `#deprecation_reason => String, nil`
+
+`GraphQL::Function` provides some help in implementing these:
+
+```ruby
+class MyFunc < GraphQL::Function
+  # Define a member of `#arguments`, just like the DSL:
+  argument :id, GraphQL::ID_TYPE
+
+  # Define documentation:
+  description "My Custom function"
+  deprecation_reason "Just an example"
+
+  type MyFuncReturnType
+  # or, define one on the fly:
+  type do
+    name "MyFuncReturnType"
+    # The returned object must implement these methods:
+    field :name, GraphQL::STRING_TYPE
+    field :count, GraphQL::INT_TYPE
   end
 end
-
-# ...
-field :name, types.String do
-  resolve(MethodCallResolver.new(:name))
-end  
 ```
 
-Or, you can generate the `resolve` Proc dynamically:
+Note that `types.` is _not_ available. Instead, you should reference GraphQL's built-in {{ "GraphQL::ScalarType" | api_doc }}s directly.
```

Comment:
These docs also don't mention the "extensibility" of the feature. I thought this might have been removed, but it appears not!

What I mean is, the ability to add additional `arguments` or whatever to the global function.